VS-280 Create a VAT intermediary (#7657)
broadinstitute/gatk: scripts/variantstore/variant_annotations_table/ReadMe.md

Creating the Variant Annotations Table

The WDL to run is:
GvsCreateVAT

And then you’ll want to run:
GvsValidateVAT
To check and validate the VAT

To run GvsCreateVAT:

> Most of the inputs are constants---like the reference, or a table schema--and dont require any additional work (and paths can be found in the example json). However for the specific data being put in the VAT, three inputs need to be created.

Reviewer: ```suggestion
Most of the inputs are constants — like the reference, or a table schema — and don't require any additional work (and paths can be found in the [example json](/scripts/variantstore/wdl/GvsCreateVAT.example.inputs.json)). However for the specific data being put in the VAT, three inputs need to be created.
```

Reviewer: (not sure if the link path is right, but it would be useful for it be a link)

> The first two of these inputs are two files—one of the file/vcf/shards you want to use for the VAT, and one of their indices. These are labelled as inputFileofFileNames and inputFileofIndexFileNames.

Reviewer: ```suggestion
The first two of these inputs are two files — the file/vcf/shards you want to use for the VAT, and their corresponding index files. These are labelled as inputFileofFileNames and inputFileofIndexFileNames.
```

You will need to know where all of the GVS extract output files live.
Example of creating the files of file names and indices:
! gsutil ls gs://prod-drc-broad/beta-release-99k-v3/beta_99k_*.vcf.gz > inputFileofFileNames.txt
! gsutil ls gs://prod-drc-broad/beta-release-99k-v3/beta_99k_*.vcf.gz.tbi > inputFileofIndexFileNames.txt

Those two created files will then need to get copied into a GCP bucket that this pipeline will have access to (eg. this bucket: gs://aou-genomics-curation-prod-processing/vat/)

The third input is the ancestry file that Lee has made and it needs to be copied into a GCP bucket that this pipeline will have access to (eg. this bucket: gs://aou-genomics-curation-prod-processing/vat/) for easy access during the workflow. This input has been labelled as the ancestry_file.

> Most of the other files are specific to where the VAT will live, like the project_id and dataset_name and the table_suffix which will name the VAT itself as vat_`table_suffix` as well as a GCP bucket location, the output_path, for the intermediary files and the VAT export in tsv form.

Reviewer: ```suggestion
Most of the other files are specific to where the VAT will live — like the `project_id`, `dataset_name` and `table_suffix` (which will name the VAT itself as vat_`table_suffix`) — as well as a GCP bucket location, the `output_path`, for the intermediary files and the VAT export in TSV form.
```

> The other inputs can be copied from where I’ve put several files (maybe we should find a better place?!? Since many are in my scratch dir for now—tho anything in my scratch dir is also living in github as it’s things like what BQ schema is needed)

Reviewer: What location are you referring to?

The provided example WDL indicates all of the needed inputs.

> Here’s a brief overview of the pipeline itself:

Reviewer: I think that the task-by-task explanation can be done either here **or** in WDL comments but it doesn't make sense for it to be in two places (too easy to go out-of-sync).  I'd lean toward WDL comments but I don't feel too strongly.

  ExtractAnAcAfFromVCF is the first step of the subworkflow and filters out bad sites, sites with too many alt alleles, duplicate variants and calculates the AC and AN values for each subpopulation
  AnnotateVCF is the second step of the subworkflow and adds Nirvana annotations to each of the variants
  PrepAnnotationJson is the final step and prepares the data to be loaded into a BQ table, and loads it into a GCP staging bucket, using the output_path

BigQueryLoadJson creates BQ tables (two temporary and one that is the VAT) and then gets everything from the staging buckets (genes and vt) and loads it into the two temp tables in BQ. Those two tables are joined together to create the VAT.

BigQuerySmokeTest checks that the number of variants in the new VAT is the same as the expected number from the many subworkflows. This will never fail the pipeline, but is helpful to validate results.

BigQueryExportVat takes the VAT and exports it by chromosome into a series of tsv files and loads it into a GCP bucket in the directory created by the output_path

Notes:

> When running this pipeline, currently I try to stay around 5k-10k shards—I have yet to successfully run more than that, but I think it’s more about nerves than anything else. At 20k shards, the shards do tend to step on each other if there are too many—not that anything fails or is wrong, but I’d rather get results for a run that I then do multiple times than wait for queued jobs.

Reviewer: ```suggestion
When running this pipeline, I'd suggest starting with 5k-10k shards at a time. At 20k shards, the shards do tend to step on each other.
```

> The VAT table is created by that query fresh each time so that there is no risk of duplicates.

Reviewer: Maybe put this note next to the query in question for additional clarity?

> HOWEVER the Genes and VT tables are not. They are cleaned up after 24 hours, but this code needs to be tweaked so that you can’t get into a state where duplicates are created here. The real question here is going to be, is there a use case that we might want to run where adding to a VAT that was created say weeks ago is beneficial, but given that calculations

Reviewer: not sure what the last sentence is trying to convey here

> Note that there are two temporary tables that are created in addition to the main VAT table

Reviewer: If you are going to include this, I would list the tables (and maybe which workflow creates them), too, but I'm not sure why this needs to be documented if they clean up themselves?  Or is there a scenario where someone might need to manually clean them up?

> To check that all of the shards have successfully made it past the first of the sub workflow steps (the most complicated and likely to fail) they will need to be annotated / transformed into json files and put here:

Reviewer: Does the person running the workflow need to do the annotation/transformation or is the workflow doing that?

There are often a fair number of failures from google in this workflow—-so far they have all been 503s. Because of the re-working of the workflow, they should not interrupt unaffected shards, but the shards that are affected will need to be collected and put into a new File of File names and re-run.
In theory you could actually just re-run the pipeline entirely and since the 503s seem to be completely random and intermittent, the same shards likely wont fail twice and you’d get everything you need, but at low efficiency and high expense.
To grab any files not in the bucket, but in the file of file names:

gsutil ls [output_path]/genes/ | awk '{print substr($0,90)}' RS='.json.gz\n'  > successful_vcf_shards.txt
gsutil cat [inputFileofFileNames] | awk '{print substr($0,41)}' RS='.vcf.gz\n' > all_vcf_shards.txt
comm -3 all_vcf_shards.txt successful_vcf_shards.txt > diff_vcf_shards.txt
awk '{print g[bucket GVS output VCF shards are in]" $0 ".vcf.gz"}' diff_vcf_shards.txt > fofn_missing_shards.txt

> (dont forget to do this for the indices as well)

Reviewer: Why would a file exist (or not) for a VCF file but not it's corresponding index?

> There is a line of code in ExtractAnAcAfFromVCF (the most expensive in $ and time task in the workflow) that can be removed because it is used to track the variants that are dropped—TODO: Rori to make using it a parameter.

Reviewer: I think this would be most useful as a ticket, instead of a note here.
